### PR TITLE
Add `_resolve_execution_config` helper function to `qnode.py`

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

The goal is to streamline the process of updating the execution configuration with non-device-specific settings. This approach consolidates the setup and validation of the execution configuration in one spot in the code, before calling `qml.execute` in `qnode.py`.

**Description of the Change:**

This refactor organizes the setup of the initial `config` by moving it into a dedicated helper function. It also includes the setup and validation code for handling the `MCMConfig`.

**Benefits:** Improves code cleaniness in `qnode.py` and allows for better testing of internal pipeline.

**Possible Drawbacks:** None

[sc-72086]

### CODE REVIEW

File: pennylane/workflow/qnode.py [hunk 1]
```diff
@@ -96,13 +99,92 @@ def _make_execution_config(
...
+        qml.devices.ExecutionConfig: resolved execution configuration
+    """
+    if (
+        device.name == "lightning.qubit"
+        and qml.metric_tensor in transform_program
+        and execution_config.gradient_method == "best"
+    ):
+        gradient_fn = qml.gradients.param_shift
+    else:
+        gradient_fn = QNode.get_gradient_fn(
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:156]: Potentially worth relying on something else so that we can deprecate `get_gradient_fn`.

File: pennylane/workflow/qnode.py [hunk 2]
```diff
@@ -96,13 +99,92 @@ def _make_execution_config(
...
+            device, execution_config.interface, execution_config.gradient_method, tape=tapes[0]
+        )[0]
+
+    if gradient_fn is qml.gradients.param_shift_cv:
+        execution_config.gradient_keyword_arguments["dev"] = device
+
+    execution_config.gradient_method = gradient_fn
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:163]: We should probably also be using `dataclasses.replace` or first copying the execution config so that we are not mutating the input to this function.

File: pennylane/workflow/qnode.py [hunk 3]
```diff
@@ -96,13 +99,92 @@ def _make_execution_config(
...
+    finite_shots = any(tape.shots for tape in tapes)
+    if not finite_shots:
+        execution_config.mcm_config.postselect_mode = None
+        if execution_config.mcm_config.mcm_method == "one-shot":
+            raise ValueError(
+                f"Cannot use the '{execution_config.mcm_config.mcm_method}' method for mid-circuit measurements with analytic mode."
+            )
+
+    if execution_config.mcm_config.mcm_method == "single-branch-statistics":
+        raise ValueError("Cannot use mcm_method='single-branch-statistics' without qml.qjit.")
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:174]: Could probably place this validation into the `_update_mcm_config` step.

File: pennylane/workflow/qnode.py (hunk 1 above)
Comment (andrijapau) [reply pennylane/workflow/qnode.py:174]: Yeah ... I tried to use the new `qml.workflow._get_gradient_fn` earlier but it caused some tests to fail. I'm taking a look at it now and trying to make it work. 😄 

The problem is with this branch in `_get_gradient_fn`,
```python
...
    if diff_method == "best":
        qn = qml.QNode(lambda: None, device, diff_method=None)
        return qml.workflow.get_best_diff_method(qn)()
...
```
Since `get_best_diff_method` has no direct access to the tape provided to `_get_gradient_fn` it incorrectly returns `backprop` when it should be `parameter-shift` for finite shots. I tried setting the tape with `qn._tape = tape` but that didn't work. 

Any ideas?

Comment (andrijapau) [reply pennylane/workflow/qnode.py:171]: I think the best way forward would be to modify the `qml.workflow._get_gradient_fn` to not rely on `qml.workflow.get_best_diff_method` and instead replace that block of code with the necessary logic from the latter,
```python
...
    if diff_method == "best":
        config = _make_execution_config(None, "best")

        if device.supports_derivatives(config, circuit=tape):
            new_config = device.preprocess(config)[1]
            return new_config.gradient_method

        if tape and any(isinstance(o, qml.operation.CV) for o in tape):
            return qml.gradients.param_shift_cv
            
        return qml.gradients.param_shift
 ...
```

File: pennylane/workflow/get_gradient_fn.py [hunk 4]
```diff
@@ -61,8 +61,16 @@ def _get_gradient_fn(
...
+        config = _make_execution_config(None, "best")
+
+        if device.supports_derivatives(config, circuit=tape):
+            new_config = device.preprocess(config)[1]
+            return new_config.gradient_method
+
+        if tape and any(isinstance(o, qml.operation.CV) for o in tape):
+            return qml.gradients.param_shift_cv
+
+        return qml.gradients.param_shift
```
Comment (andrijapau) [inline pennylane/workflow/get_gradient_fn.py:73]: Moved this logic from `qml.workflow.get_best_diff_method`. No longer need to create a dummy `QNode`; which prevented me using this helper function in `qnode.py`

File: pennylane/workflow/qnode.py (hunk 1 above)
Comment (lillian542) [reply pennylane/workflow/qnode.py]: I think you just need to do something like pass shots from the tape in when calling the output of `get_best_diff_method` inside `_get_gradient_fn` and that should resolve. 

I.e. something along the lines of `return qml.workflow.get_best_diff_method(qn)(shots=tape.shots)`

Comment (andrijapau) [reply pennylane/workflow/qnode.py]: Interesting, I can try that.

Although, this might actually be an improvement to `_get_gradient_fn` since `get_best_diff_method` is designed to be user-friendly - returning strings rather than transform functions. Seems kind of weird for an internal developer function to rely on a user-designed function? 

By extracting the relevant code from `get_best_diff_method`, we could simply return the transform function directly.

Comment (lillian542) [reply pennylane/workflow/qnode.py]: I can see that it might be nice if `_get_gradient_fn` returned a function instead of a string. I don't have a strong opinion either way - since we are allowing users to pass in `diff_method`, we are going to have a mix of strings and functions to define our gradients up to some point, and I _think_ at least in our current implementation we still have a mix at this point.

File: pennylane/workflow/execution.py [hunk 5]
```diff
@@ -105,7 +105,14 @@
...
 
 def _use_tensorflow_autograph():
-    import tensorflow as tf
+    """Checks if TensorFlow is in graph mode, allowing Autograph for optimized execution"""
+    try:  # pragma: no cover
+        import tensorflow as tf
+    except ImportError as e:  # pragma: no cover
+        raise qml.QuantumFunctionError(  # pragma: no cover
+            "tensorflow not found. Please install the latest "  # pragma: no cover
+            "version of tensorflow to enable the 'tensorflow' interface."  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/workflow/execution.py:114]: I see what you mean but I was just thinking of the edge case where the user doesn't have `tensorflow` installed in their environment and they choose to use `interface="tf"`. I think having a response akin to the `jax` error would be helpful. 

> I don't think we're compatible with the latest version of tensorflow.

Good point, I can update the error message to,
```python
raise qml.QuantumFunctionError(  # pragma: no cover
            "tensorflow not found. Please install the latest "  # pragma: no cover
            "version of tensorflow supported by Pennylane " # pragma: no cover
            "to enable the 'tensorflow' interface."  # pragma: no cover
        ) from e  # pragma: no cover
```

File: pennylane/workflow/qnode.py [hunk 6]
```diff
@@ -96,13 +100,96 @@ def _make_execution_config(
...
+        transform_program (TransformProgram): a program of transformations to be applied to the tapes
+
+    Returns:
+        qml.devices.ExecutionConfig: resolved execution configuration
+    """
+    updated_values = {}
+    updated_values["gradient_keyword_arguments"] = dict(execution_config.gradient_keyword_arguments)
+
+    if (
+        device.name == "lightning.qubit"
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:165]:
```suggestion
        "lightning" in device.name
```
Potentially out of scope of this PR, but looking over this I realized the same bug might apply to kokkos and gpu too.

File: pennylane/workflow/qnode.py [hunk 7]
```diff
@@ -874,73 +957,52 @@ def _execution_component(self, args: tuple, kwargs: dict) -> qml.typing.Result:
...
 
-        if mcm_config.mcm_method == "single-branch-statistics":
-            raise ValueError("Cannot use mcm_method='single-branch-statistics' without qml.qjit.")
+        config = _make_execution_config(self, self.diff_method, mcm_config=mcm_config)
+
+        # Mapping numpy to None here because `qml.execute` will map None back into
+        # numpy. If we do not do this, numpy will become autograd in `qml.execute`.
+        # If the user specified interface="numpy", it would've already been converted to
+        # "autograd", and it wouldn't be affected.
+        _interface_user_input = None if self.interface == "numpy" else self.interface
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:970]: So I think we'll be able to clean this component up if we start using an enum for the interface designation?

File: tests/transforms/test_add_noise.py [hunk 8]
```diff
@@ -221,7 +221,7 @@ def test_add_noise_dev(self, dev_name):
         new_program, _ = new_dev.preprocess()
         [tape], _ = new_program([in_tape])
         res_with_noise = qml.execute(
-            [in_tape], new_dev, qml.gradients, transform_program=new_program
+            [in_tape], new_dev, qml.gradients.param_shift, transform_program=new_program
```
Comment (albi3ro) [inline tests/transforms/test_add_noise.py:224]: I think this supports my desire to mode gradient method validation inside `qml.execute`... 😬  How did that even work 🫠

File: pennylane/workflow/get_gradient_fn.py [hunk 9]
```diff
@@ -26,7 +27,8 @@
 )
 
 
-# pylint: disable=too-many-return-statements, unsupported-binary-operation
+# pylint: disable=too-many-return-statements, unsupported-binary-operation, inconsistent-return-statements
+@debug_logger
```
Comment (albi3ro) [inline pennylane/workflow/get_gradient_fn.py:31]: Potentially out of scope, but we should be able to make `tape` a mandatory argument now, since we aren't needing to worry about backward compatibility anymore.

File: pennylane/workflow/get_gradient_fn.py [hunk 10]
```diff
@@ -86,7 +96,3 @@ def _get_gradient_fn(
 
     if isinstance(diff_method, TransformDispatcher):
         return diff_method
-
-    raise qml.QuantumFunctionError(
-        f"Differentiation method {diff_method} must be a gradient transform or a string."
-    )
```
Comment (albi3ro) [inline pennylane/workflow/get_gradient_fn.py:92]: Potentially we can move the `if isinstance(diff_method, str)` block below:

```
        raise qml.QuantumFunctionError(
            f"Differentiation method {diff_method} not recognized. Allowed "
            f"options are {tuple(get_args(SupportedDiffMethods))}."
        )
```

That way we aren't left with a hanging ending to the function.

File: pennylane/workflow/get_gradient_fn.py
Comment (albi3ro) [inline pennylane/workflow/get_gradient_fn.py:1]: Since the whole purpose of `_get_gradient_fn` is to be a step inside of `_resolve_execution_config`, we can update the call signature to:
```
def _resolve_diff_method(initial_config, device, tape):
```

Then we can determine whether or not the device supports the diff method for the full initial config, and not just the temporary version created for the time being.  

We could even update it to be:
```
def _resolve_diff_method(initial_config, device, tape) -> "ExecutionConfig":
```
Where it updates does the device setup, and if the device doesn't support the diff method, it fills in `gradient_method`.

File: tests/transforms/test_add_noise.py (hunk 8 above)
Comment (andrijapau) [reply tests/transforms/test_add_noise.py:224]: I know 🫨 !

File: pennylane/workflow/qnode.py [hunk 11]
```diff
@@ -96,13 +100,96 @@ def _make_execution_config(
...
-        gradient_method=_gradient_method,
+        gradient_keyword_arguments=gradient_kwargs,
+        gradient_method=diff_method,
         grad_on_execution=grad_on_execution,
         use_device_jacobian_product=execute_kwargs.get("device_vjp", False),
         mcm_config=mcm_config or qml.devices.MCMConfig(),
     )
 
 
+def _update_mcm_config(mcm_config: "qml.devices.MCMConfig", interface: str, finite_shots: bool):
```
[the remark this answers is not in the document]
Comment (andrijapau) [reply pennylane/workflow/qnode.py:111]: On it 😎

File: pennylane/workflow/execution.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/workflow/execution.py]: Is _this_ the error that is raised if someone does that? I would hope we hit a user-facing error sooner than that (pretty much immediately), but I guess this is quite early in the pipeline. Either way, suggestion above looks good to me!

File: pennylane/workflow/qnode.py [hunk 12]
```diff
@@ -874,73 +964,52 @@ def _execution_component(self, args: tuple, kwargs: dict) -> qml.typing.Result:
...
-        if mcm_config.mcm_method == "single-branch-statistics":
-            raise ValueError("Cannot use mcm_method='single-branch-statistics' without qml.qjit.")
+        config = _make_execution_config(self, self.diff_method, mcm_config=mcm_config)
+
+        # Mapping numpy to None here because `qml.execute` will map None back into
+        # numpy. If we do not do this, numpy will become autograd in `qml.execute`.
+        # If the user specified interface="numpy", it would've already been converted to
+        # "autograd", and it wouldn't be affected.
+        _interface_user_input = None if self.interface == "numpy" else self.interface
+        config.interface = _interface_user_input
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:978]: Could we move this step into `_make_execution_config`?

File: pennylane/workflow/get_gradient_fn.py
Comment (andrijapau) [reply pennylane/workflow/get_gradient_fn.py:1]: @albi3ro, addressed in the newest commit!

File: pennylane/workflow/resolve_diff_method.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    diff_method = initial_config.gradient_method
+    updated_values = {"gradient_method": diff_method}
+
+    if diff_method is None:
+        return initial_config
+
+    if device.supports_derivatives(initial_config, circuit=tape):
+        _, new_config = device.preprocess(initial_config)
+        updated_values["gradient_method"] = new_config.gradient_method
+        return replace(initial_config, **updated_values)
```
Comment (albi3ro) [inline pennylane/workflow/resolve_diff_method.py:53]:
```suggestion
        return new_config
```

**Review by albi3ro — APPROVED** [review]
❤️

**Review by astralcai — APPROVED** [review]
LGTM!
